src/parseAndDecorate.ts

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 Google LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     https://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+'use strict';
+import path = require('path');
+import { JSONPath, visit } from 'jsonc-parser';
+import * as vscode from 'vscode';
+import XRegExp = require('xregexp');
+
+export const diagnostics = vscode.languages.createDiagnosticCollection("arb");
+
+const argDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#ff6f00'
+	},
+	dark: {
+		color: '#fff9c4'
+	}
+});
+const selectDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#6a1b9a'
+	},
+	dark: {
+		color: '#ce93d8'
+	}
+});
+const pluralDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#0277bd'
+	},
+	dark: {
+		color: '#b3e5fc'
+	}
+});
+
+const selectRegex = /^(\w+\s*,\s*(?:select|gender)\s*,(?:\s*\w+\{.*\})*)$/;
+const pluralRegex = /^(\w+\s*,\s*plural\s*,(?:\s*\w+\{.*\})*)$/;
+const argNameRegex = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/;
+const keyNameRegex = /^[a-zA-Z][a-zA-Z_0-9]*$/;
+
+class Literal {
+	constructor(
+		public value: string,
+		public start: number,
+		public end: number
+	) { }
+
+	public toString = (): string => {
+		return `Literal(${this.value},${this.start},${this.end})`;
+	};
+}
+
+export function parseAndDecorate(editor: vscode.TextEditor) {
+	// Prefill decorations map to avoid having old decoration hanging around
+	let decorationsMap = new Map<vscode.TextEditorDecorationType, vscode.Range[]>([
+		[argDecoration, []],
+		[selectDecoration, []],
+		[pluralDecoration, []],
+	]);
+	// Clear diagnostics
+	diagnostics.set(editor.document.uri, []);
```

Comment:
Do you need this? I would expect line 151 to take care of this (since it should set an empty array if no new diagnostics are found). Setting it twice could result in flickering because VS Code's UI is in another process, so it might remove all the diagnostics and then put them back as the user is typing.